packages/govuk-frontend/src/govuk/components/input/template.njk

```diff
@@ -39,30 +60,17 @@
   }) | trim | indent(2) }}
 {% endif %}
 
-  {%- if params.prefix or params.suffix %}<div class="govuk-input__wrapper">{% endif -%}
-
-  {%- if params.prefix.text or params.prefix.html -%}
-    <div class="govuk-input__prefix {%- if params.prefix.classes %} {{ params.prefix.classes }}{% endif %}" aria-hidden="true" {%- for attribute, value in params.prefix.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>
-      {{- params.prefix.html | safe if params.prefix.html else params.prefix.text -}}
-    </div>
-  {% endif -%}
-
-  <input class="govuk-input {%- if params.classes %} {{ params.classes }}{% endif %} {%- if params.errorMessage %} govuk-input--error{% endif %}" id="{{ params.id }}" name="{{ params.name }}" type="{{ params.type | default("text", true) }}"
-  {%- if (params.spellcheck === false) or (params.spellcheck === true) %} spellcheck="{{ params.spellcheck }}"{% endif %}
-  {%- if params.value %} value="{{ params.value }}"{% endif %}
-  {%- if params.disabled %} disabled{% endif %}
-  {%- if describedBy %} aria-describedby="{{ describedBy }}"{% endif %}
-  {%- if params.autocomplete %} autocomplete="{{ params.autocomplete }}"{% endif %}
-  {%- if params.pattern %} pattern="{{ params.pattern }}"{% endif %}
-  {%- if params.inputmode %} inputmode="{{ params.inputmode }}"{% endif %}
-  {%- for attribute, value in params.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>
-
-  {%- if params.suffix.text or params.suffix.html -%}
-    <div class="govuk-input__suffix {%- if params.suffix.classes %} {{ params.suffix.classes }}{% endif %}" aria-hidden="true" {%- for attribute, value in params.suffix.attributes %} {{ attribute }}="{{ value }}"{% endfor %}>
-      {{- params.suffix.html | safe if params.suffix.html else params.suffix.text -}}
-    </div>
-  {% endif -%}
-
-{%- if params.prefix or params.suffix %}</div>{% endif %}
-
+{%- if hasPrefix or hasSuffix %}
+  <div class="govuk-input__wrapper">
+    {% if hasPrefix %}
+      {{- _affixItem(params.prefix, "prefix") | indent(2, true) }}
+    {% endif %}
+    {{ _inputElement(params) }}
+    {% if hasSuffix %}
+      {{- _affixItem(params.suffix, "suffix") | indent(2, true) }}
+    {% endif %}
+  </div>
+{% else %}
+  {{ _inputElement(params) }}
+{% endif %}
```

Comment:
Linking how the encapsulation makes the structure very apparent and easier to follow than the `{%- if}` triggering the rendering of opening/closing tags 😍

Reply:
Thanks, same! It helped me understand it too